[MM-47566] Show upcoming invoice in subscription screen if present

nickmisasi opened the pull request:
#### Summary
With yearly subscriptions we want to be communicative to the user on what their upcoming invoice looks like. This PR exposes the upcoming invoice where the "last invoice" card would usually be if there is an upcoming invoice in the Subscription object. 

The card looks like this:
![image](https://user-images.githubusercontent.com/12176405/203634458-8ca37f05-d984-4d4d-8aea-9628e690ead0.png)
Clicking "View Invoice" will display the invoice PDF Preview modal, which gives the option to download (this change also takes effect for "last invoice")

I fixed up a bug where the line items would always just be `0` for monthly subscriptions since they were looking for `partial`/`full` line items that no longer exist.

##### QA
We will not be able to test this in a Spinwick until https://github.com/mattermost/customer-web-server/pull/1000 lands in master

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-47566

#### Related Pull Requests
https://github.com/mattermost/mattermost-server/pull/21717
https://github.com/mattermost/customer-web-server/pull/1000
#### Screenshots
![image](https://user-images.githubusercontent.com/12176405/203634891-60fd3802-bfb0-402e-9c8e-a0ff76de6927.png)


#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
None
```

neallred commented on components/admin_console/billing/billing_summary/billing_summary.tsx:
```diff
@@ -128,6 +133,16 @@ export const getPaymentStatus = (status: string) => {
+                    id='admin.billing.subscriptions.billing_summary.lastInvoice.upcoming'
+                    defaultMessage='Upcoming'
+                />
+                <i className='icon icon-check-circle-outline'/>
```
Nit: Prefer
```
import CheckCircleOutlineIcon from '@mattermost/compass-icons/components';
...
<CheckCircleOutlineIcon />
```

neallred commented on components/admin_console/billing/billing_summary/billing_summary.tsx:
```diff
@@ -172,7 +209,7 @@ export const lastInvoiceInfo = (invoice: any, product: any, fullCharges: any, pa
                 >
                     <div className='BillingSummary__lastInvoice-chargeDescription'>
-                        {product.billing_schema === BillingSchemes.FLAT_FEE ?
+                        {product?.billing_scheme === BillingSchemes.FLAT_FEE ?
```
Oh interesting. I wonder why TypeScript didn't catch this before.

nickmisasi replied:
The product was a function arg with `any` type. I changed it to a component and gave it a proper type

neallred commented on components/admin_console/billing/billing_summary/billing_summary.tsx:
```diff
@@ -202,62 +239,64 @@ export const lastInvoiceInfo = (invoice: any, product: any, fullCharges: any, pa
-                            id='BillingSubscriptions__seatOverageTooltip'
-                            className='BillingSubscriptions__tooltip BillingSubscriptions__tooltip-right'
-                            positionLeft={390}
+            {partialCharges.length ?
```
Good catch! Could we do `Boolean(partialCharges.length) && ...` instead of `{partialCharges.length ? ... : null`. It's a lot of lines to scan to see that what we're doing is not making two large conditional render branches but safeguarding against displaying `0`.

neallred commented on components/admin_console/billing/billing_summary/index.tsx:
```diff
@@ -41,12 +41,35 @@ const BillingSummary = ({isFreeTrial, daysLeftOnTrial, onUpgradeMattermostCloud}
+            />
+        );
+    } else if (subscription?.upcoming_invoice) {
+        const invoice = subscription!.upcoming_invoice;
```
Nit: I don't think the `!` assertion is needed here because the else if already checks it. If I remove it locally my editor doesn't complain.

neallred commented on components/admin_console/billing/billing_summary/billing_summary.tsx:
```diff
@@ -294,18 +333,22 @@ export const lastInvoiceInfo = (invoice: any, product: any, fullCharges: any, pa
-                    className='BillingSummary__lastInvoice-downloadButton'
+                <button
+                    onClick={() => dispatch(openModal({
+                        modalId: ModalIdentifiers.CLOUD_INVOICE_PREVIEW,
```
:trophy:

neallred reviewed (approved):
LGTM! A couple nits, but nothing blocking.

marianunez commented on components/admin_console/billing/billing_summary/billing_summary.tsx:
```diff
@@ -15,6 +18,9 @@ import EmptyBillingHistorySvg from 'components/common/svg_images_components/empt
 import {trackEvent} from 'actions/telemetry_actions';
 
 import {Client4} from 'mattermost-redux/client';
+import {Invoice, InvoiceLineItem, Product} from '@mattermost/types/cloud';
+import {openModal} from 'actions/views/modals';
+import CloudInvoicePreview from 'components/cloud_invoice_preview';
```
nit: can we add these in the groups their related groups `components`, `actions` to keep this ordered?

marianunez reviewed (commented):
Thanks @nickmisasi! Code looks good but would like for @johndavidlugtu and @annelieseklein to take a look at the copies and UX to see if we need to add any additional context on the users added there.

marianunez commented:
> #### Screenshots
> ![image](https://user-images.githubusercontent.com/12176405/203634891-60fd3802-bfb0-402e-9c8e-a0ff76de6927.png)

- Will the upcoming date be at the end of the annual subscription?
- How will it handle when it has more than 5 lines or so that it overflows that space?
- Is each line the accumulated monthly true-up? If so, should we include the month in some form in the line? Real estate is tough but listing just the count in multiple lines feels incomplete of when that charge happened.

marianunez reviewed (commented):
Can we include smoke testing the latest invoice for monthly subscriptions as part of the QA steps for this ? cc/ @yasserfaraazkhan

annelieseklein commented:
1. I don't think we need the "upcoming" in grey in the right
2. Can we make the "View Invoice" button be full width with "see billing history" centered below it

nickmisasi commented:
@annelieseklein 
Removed the grey "upcoming" on the right.
Centered the "View Invoice" and "See billing history" buttons


@marianunez @johndavidlugtu 
I also added a truncation mechanism - if there are more than 5 line items, the component will display 5 of them, and then list `And {count remaining} more`.

Let me know what you think of the way I've got it here:
![image](https://user-images.githubusercontent.com/12176405/205986349-61b2abb5-902b-47b0-afa1-6d37c003b2fa.png)

nickmisasi commented:
@marianunez 
>Will the upcoming date be at the end of the annual subscription?
It will match the period_start of the upcoming invoice, which should be the date it's due in this scenario. 

>How will it handle when it has more than 5 lines or so that it overflows that space?
See https://github.com/mattermost/mattermost-webapp/pull/11638#issuecomment-1339759370 for my proposal on overflow.


    How will it handle when it has more than 5 lines or so that it overflows that space?
    Is each line the accumulated monthly true-up? If so, should we include the month in some form in the line? Real estate is tough but listing just the count in multiple lines feels incomplete of when that charge happened.
    
>Is each line the accumulated monthly true-up? If so, should we include the month in some form in the line? Real estate is tough but listing just the count in multiple lines feels incomplete of when that charge happened.

It's not an accumulated monthly thing. It's each line item. My thinking here is that this just shows at a glance the proration, and if you're curious about what each line actually means you can open the full invoice which is more descriptive. Otherwise, it's a lot more logic on both back and front-end to get everything to display nicely

annelieseklein commented:
If we do truncate, the user should be able to click to "View All". which would maybe open a scrollable modal

nickmisasi commented:
@annelieseklein Given that the "View Invoice" button displays all invoice items (ie, not truncated), does that suffice? We don't have any existing components that I can plug in to do what you're mentioning so we'd need designs and more work. 

I could even make the `And {count} more items` clickable so it opens the same modal the View Invoice button does

annelieseklein commented:
ah true, yeah all good then

annelieseklein commented:
+1 to making and more items clickable tho

annelieseklein commented:
Looking back at the invoice, we actually have a green text in the right when it's paid, so i think it'd be okay to keep the grey text but have it say "Unpaid" instead of upcoming

annelieseklein commented:
ah i see. so the invoice isn't ready to be paid yet?

annelieseklein commented:
gotcha ok, thinking we could have grey text say "Pending" potentially. or something similar to denote in-progress.

nickmisasi commented:
@annelieseklein sounds good:

![image](https://user-images.githubusercontent.com/12176405/205990804-60accf0a-4ba2-44fd-a578-e93407bf6f60.png)


The "And 4 more items" is clickable, and will turn to cursor pointer/undlerline when hovered.

nickmisasi commented:
@annelieseklein 

![image](https://user-images.githubusercontent.com/12176405/205998720-35434bfd-0021-473e-b0e4-ec8702b59320.png)

lmk if you want me to adjust the font weight. Otherwise, if you could approve that'd be great!

marianunez commented:
> he "And 4 more items" is clickable, and will turn to cursor pointer/undlerline when hovered.

What would the link open? Also, could we get a server as part of the bug bash to test this out?

nickmisasi commented:
>What would the link open? Also, could we get a server as part of the bug bash to test this out?

@marianunez it opens the invoice preview modal just like the "View Invoice" button does

johndavidlugtu commented:
+1 on the changes. Agreed that "And x more items" should open the same modal as "View Invoice"

mattermod commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/272501) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/272501)
